aws/src/test/java/org/apache/iceberg/aws/s3/TestFlakyS3InputStream.java

```diff
@@ -0,0 +1,194 @@
+        .isInstanceOf(exception.getClass())
+        .hasMessage(exception.getMessage());
+  }
+
+  @ParameterizedTest
+  @MethodSource("nonRetryableExceptions")
+  public void testReadWithFuzzyStreamNonRetryableException(IOException exception) {
+    assertThatThrownBy(
+            () -> testRead(fuzzyStreamClient(new AtomicInteger(3), exception), DATA_SIZE))
+        .isInstanceOf(exception.getClass())
+        .hasMessage(exception.getMessage());
+  }
+
+  @Override
+  protected void testRead(S3Client s3, int dataSize) throws Exception {
+    testRead(s3, DATA_SIZE, 4, SEEK_SIZE, SEEK_NEW_POSITION);
+  }
+
+  private static Stream<Arguments> retryableExceptions() {
+    return Stream.of(
+        Arguments.of(
+            new SocketTimeoutException("socket timeout exception"),
+            new SSLException("some ssl exception")));
+  }
+
+  private static Stream<Arguments> nonRetryableExceptions() {
+    return Stream.of(Arguments.of(new IOException("some generic non-retryable IO exception")));
+  }
+
+  private S3ClientWrapper fuzzyStreamClient(AtomicInteger counter, IOException failure) {
+    S3ClientWrapper fuzzyClient = spy(new S3ClientWrapper(s3Client()));
+    doAnswer(invocation -> new FlakyInputStream(invocation.callRealMethod(), counter, failure))
+        .when(fuzzyClient)
+        .getObject(any(GetObjectRequest.class), any(ResponseTransformer.class));
+    return fuzzyClient;
+  }
+
+  /** Wrapper for S3 client, used to mock the final class DefaultS3Client */
+  public static class S3ClientWrapper implements S3Client {
+
+    private final S3Client delegate;
+
+    public S3ClientWrapper(S3Client delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public String serviceName() {
+      return delegate.serviceName();
+    }
+
+    @Override
+    public void close() {
+      delegate.close();
+    }
+
+    @Override
+    public <ReturnT> ReturnT getObject(
+        GetObjectRequest getObjectRequest,
+        ResponseTransformer<GetObjectResponse, ReturnT> responseTransformer)
+        throws AwsServiceException, SdkClientException {
+      return delegate.getObject(getObjectRequest, responseTransformer);
+    }
+
+    @Override
+    public HeadObjectResponse headObject(HeadObjectRequest headObjectRequest)
+        throws AwsServiceException, SdkClientException {
+      return delegate.headObject(headObjectRequest);
+    }
+
+    @Override
+    public PutObjectResponse putObject(PutObjectRequest putObjectRequest, RequestBody requestBody)
+        throws AwsServiceException, SdkClientException {
+      return delegate.putObject(putObjectRequest, requestBody);
+    }
+
+    @Override
+    public CreateBucketResponse createBucket(CreateBucketRequest createBucketRequest)
+        throws AwsServiceException, SdkClientException {
+      return delegate.createBucket(createBucketRequest);
+    }
+  }
+
+  static class FlakyInputStream extends InputStream {
```

Comment:
I slimmed down this input stream implementation so that we only implement read/close that's required for the tests.